doc/advanced-configuration/listen.md

```diff
@@ -0,0 +1,738 @@
+A reason to do so could be reduced traffic - systems with many components could return very long disco responses.
+Also, some deployments would like to avoid revealing some services; not because it is a security threat (this method does not prevent clients from communicating with hidden components), but rather because they are not meant to interact with clients directly (e.g. helper components for other components).
+
+#### `listen.service.conflict_behaviour`
+* **Syntax:** string, one of: `"disconnect"`, `"kick_old"`
+* **Default:** `"disconnect"`
+* **Example:** `conflict_behaviour = "kick_old"`
+
+By default, when a component tries to connect and a registration conflict occures, the connection is dropped with the following error:
+
+```xml
+<stream:error>
+  <conflict xmlns='urn:ietf:params:xml:ns:xmpp-streams'/>
+</stream:error>
+</stream:stream>
+```
+
+It makes implementing the reconnection logic difficult, because the old connection would not allow any other connections.
+By setting this option to `kick_old`, we drop any old connections registered at the same host before accepting new ones.
+
+### Custom extension to the protocol
+
+In order to register a component for all virtual hosts served by the server (see `hosts` in the `general` section), the component must add the attribute `is_subdomain="true"` to the opening stream element.
+This maybe helpful if someone wants to have a single instance of a component serving multiple virtual hosts.
+The `is_subdomain` attribute is optional and the default behaviour is as described in the XEP.
+
+### Service listener example
+
+The following section configures a service listener, accepting connections from external components.
+The IP address is limited to loopback to prevent connections from different hosts.
+All components are allowed to connect, but they need to provide the password.
+The shaper named `fast` needs to be defined in the `shaper` section.
+
+```toml
+[[listen.service]]
+  port = 8888
+  access = "all"
+  shaper_rule = "fast"
+  ip_address = "127.0.0.1"
+  password = "secret"
+```
+
+## HTTP-based services: `[[listen.http]]`
+
+Manages all HTTP-based services, such as BOSH (HTTP long-polling), WebSocket and REST.
+It uses the [Cowboy](https://ninenines.eu/docs/en/cowboy/2.6/manual) web server.
+Recommended port number: 5280 for BOSH/WS.
+
+There are following options for each of the HTTP listeners:
+
+#### `listen.http.handlers`
+* **Syntax:** each handler is specified in a subsection starting with `[[listen.http.handlers.type]]` where `type` is one of the allowed handler types, handling different connection types, e.g.
+
+    * `mod_bosh` - for [BOSH](https://xmpp.org/extensions/xep-0124.html) connections,
+    * `mod_websockets` - for [WebSocket](https://tools.ietf.org/html/rfc6455) connections,
+    * `mongoose_api_*`, `mongoose_client_api_*`, ... - for REST API.
```

Comment:
If we don't list all the handlers here I'd like to see a ref added to be able to check them quickly

Reply:
The point is that the user does not need to know them all, not supposed to change them one by one. That's the point when I wanted to avoid going too deeply into the low level hardcoded modules.